designs/0003-parallel_autodiff.md

```diff
@@ -0,0 +1,701 @@
+    UnaryFunction f)
+    -> std::vector<decltype(f(*first))>;
+
+template <class InputIt, class T, class BinaryFunction>
+constexpr T parallel_reduce_sum(InputIt first, InputIt last, T init,
+    BinaryFunction f)
+    -> T;
+```
+
+The function signatures follow C++ standard library conventions. The
+It are iterators which are assumed to allow for random access, the
+function objects are passed by value and these are expected to have
+signature as explained below. How these functions are exposed in the
+Stan language is left unspecified for the moment. Certainly closures
+would be very helpful for this, but this would not be a requirement.
+
+## Parallel Map
+
+A dereferenced iterator from the InputIt sequence is passed into the
+UnaryFunction function object. The function is expected to return the
+same type of object for all evaluations in the sequence (except void
+would not be allowed).
+
+The parallel map facility is very similar to the existing `map_rect`
+with a few changes:
+
+- The input sequence are iterators which can point to arbitrary
+  containers (like a std vector of whatever) or these can be mapped to
+  count iterators so that the function to be evaluated gets integer
+  indices.
+- The output from the parallel map will always be a std vector and
+  each element is the result of the respective function evaluation.
+
+In practice, the parallel map will map the input iterator sequence
+onto an index set which defines the output std vector. Doing so makes
+every function evaluation per element of the input sequence fully
+*independent* of any other evaluation. Note that the function only
+requires that each element out is of the same c++ type, but it's
+actual shape is allowed to vary between elements. A direct export of
+the function to the Stan language may thus need additional
+considerations.
+
+Here is a simple unit test which demonstrates the use of the
+`parallel_map`:
+
+```c++
+// we evaluate the following binary function over a large number of
+// inputs
+
+// fun1(x, y) = (x^2 * y) + (3 * y^2)
+struct fun1 {
+  template <typename T>
+  inline T operator()(const Matrix<T, Dynamic, 1>& x) const {
+    return x(0) * x(0) * x(1) + 3.0 * x(1) * x(1);
+  }
+};
+
+TEST(Base, parallel_for) {
+  const int num_jobs = 1000;
+  // Here we use simple integer indices for the iterators
+  typedef boost::counting_iterator<int> count_iter;
+
+  vector_d x_ref_2(num_jobs);
+
+  for (int i = 0; i < num_jobs; ++i) {
+    x_ref_2(i) = 7 + i;
+  }
+
+  // UnaryFunction which returns a Eigen column vector of length 1
+  auto loop_fun = [&](int i) -> vector_d {
+    fun1 f;
+    vector_d res(1);
+    vector_d iarg(2);
+    iarg << 5, x_ref_2(i);
+    res(0) = f(iarg);
+    return res;
+  };
+
+  // the concatenate_row takes the returned std::vector<vector_d>
+  // and appends them horizontally together
+  vector_d parallel_result = stan::math::concatenate_row(
+      stan::math::parallel_map(count_iter(0), count_iter(num_jobs), loop_fun));
```

Comment:
So this will be something similar to [std::for_each](https://en.cppreference.com/w/cpp/algorithm/for_each) in C++17 right and could potentially be replaced once (if) we decide to move to C++17?

Reply:
Ah, you mention C++17 in the rest of the paragraphs, ignore me :)

Reply:
C++17 parallel is nice... but clang and gcc anyway will use the TBB as the foundation to build up C++17 parallelism. To me the TBB is the better API to go with as it is more advanced. The C++17 feels like a v1.0 things with many things missing.

But to your comment: yeah the parallel_map is like a parallel for_each; with more control over the execution (for example, the blocked range concept of the TBB is nice, I think).